### PR TITLE
[admin] Added JS widget for notififcations

### LINKED ISSUE #4: [admin] Add a JS widget to browse notifications
- [x] The widget should support infinite scrolling.
- [x] User should be able to read description of notifications.
- [x] Opening a notification should mark them as read

**Project Idea**
> Create a javscript widget for the admin which allows to view all notification and expand their details by clicking on it.
This widget shall support infinite scrolling (we have a nice implementation in django-ipam which can be reused, either partially or totally) and expand the markdown into HTML.
The operation of opening a notification detail shall flag it as read.
This feature will require adding API endpoints to get notification list and details.

Should be done after #3

### PR DESCRIPTION
**Pull request by pandafy**
TODOS

- [x] Show loader when request to server is made.
- [x] Setup ~JSLint~ JSHint 
- [ ] Add feature for #9 

Closes #4 

Hnet-image (9)

### CODE REVIEW

**Comment by pandafy** [comment]
Thanks @nemesisdesign , I will make these changes in the UI. 

> How are you flagging notifications as read now?

Through AJAX calls to API, there is an endpoint to flag a single notification as read. 
If you are asking about visually, read notifications have a faded background color.. 

> Is the bell icon updating the count?

It will get updated after a page reload for now. After implementing web sockets we can update it instantly. 

I'll get this PR to an acceptable level and will start working on web sockets simultaneously. 😄

File: openwisp_notifications/static/openwisp_notifications/js/notifications.js [hunk 1]
```diff
@@ -5,9 +5,12 @@ django.jQuery(function($) {
         new_desc;
     window.gettext = window.gettext || function (t) { return t };
     if (data.url) {
+        print("ASdasd")
```
Comment (Vivekrajput20) [inline openwisp_notifications/static/openwisp_notifications/js/notifications.js:8]: remove this line

**Comment by Vivekrajput20** [comment]
@pandafy I think you should setup eslint instead of JSLint. Eslint is the most recent and is highly extensible. It also has a plugin for django [eslint-plugin-django](https://www.npmjs.com/package/eslint-plugin-django)
Also we have already integrated it in some of our repositories like [openwisp-wifi-login-pages](https://github.com/openwisp/openwisp-wifi-login-pages), [netjsongraph.js](https://github.com/openwisp/netjsongraph.js). So it will be easier for you to setup it.

**Comment by Vivekrajput20** [comment]
Regarding the web sockets implementation,  are you planning to use [django-channels](https://channels.readthedocs.io/en/latest/) @pandafy ?
@nemesisdesign  do we use [gunicorn](https://gunicorn.org/) server for deploying it on production or any other server?

**Comment by pandafy** [comment]
> Regarding the web sockets implementation, are you planning to use [django-channels](https://channels.readthedocs.io/en/latest/) @pandafy ?

Yes @Vivekrajput20 .

**Comment by pandafy** [comment]
@Vivekrajput20 I have configured JSHint to keep JS linting simple, what do you say?  Should I move to ESLint?

File: openwisp_notifications/templatetags/notification_tags.py [hunk 2]
```diff
@@ -32,4 +32,23 @@ def unread_notifications(context):
     return output
 
 
+def notification_widget(context):
+    return format_html(
+        '''
+        <div class="notification-dropdown hide">
+            <div class="filters">
+                <span class="btn" id="mark-all-read">Mark all read</span>
```
Comment (Vivekrajput20) [inline openwisp_notifications/templatetags/notification_tags.py:40]: disable onClick event when these buttons are disabled. Currently theses buttons send API request onClick even if they are disabled.
Also either use `<button>` tag or add role and tabindex attributes to them

File: openwisp_notifications/static/openwisp_notifications/css/notification_widget.css [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+.hide {
+  display: none;
+}
+
+.info-url {
+  text-align: right;
+}
+
+.accordion {
+  overflow-y: scroll;
```
Comment (Vivekrajput20) [inline openwisp_notifications/static/openwisp_notifications/css/notification_widget.css:23]: make it auto

File: openwisp_notifications/templatetags/notification_tags.py [hunk 4]
```diff
@@ -32,4 +32,23 @@ def unread_notifications(context):
...
+                <span class="btn" id="show-unread">Show unread only</span>
+            </div>
+            <div class="accordion"></div>
+            <div class="no-notifications hide">
+                <p>No notifications founds.</p>
```
Comment (Vivekrajput20) [inline openwisp_notifications/templatetags/notification_tags.py:45]: Change this text to "No new notification"

File: openwisp_notifications/static/openwisp_notifications/css/notification_widget.css [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+  border: none;
+  outline: none;
+  transition: 0.4s;
+  background-color: #7a7a7a;
+  color: white;
+  font-weight: 600;
+}
+
+.accordion-panel {
+  padding: 0 18px;
```
Comment (Vivekrajput20) [inline openwisp_notifications/static/openwisp_notifications/css/notification_widget.css:42]: add some top padding

**Comment by Vivekrajput20** [comment]
good work @pandafy 
these are some of the things I noticed and needs to be added/fixed

- Infinite scroll does not work until the user clicks on any of the notification. 
- mark all read and mark read does not update the notification count on the bell icon
- If the last notification is clicked the content is opened but content is out of the window and not visible
- Current background colors for read and unread notification are not very intuitive, I think we can keep white background for read notifications and give background color to unread notifications only. What do you think @pandafy  @nemesisdesign?

File: openwisp_notifications/static/openwisp_notifications/css/notifications.css [hunk 6]
```diff
@@ -19,10 +19,10 @@
...
 
 .bell-icon {
-    height: 14px;
-    vertical-align: middle;
-    position: relative;
-    top: -1px;
+  height: 14px;
+  vertical-align: middle;
+  position: relative;
+  top: -1px;
```
Comment (Vivekrajput20) [inline openwisp_notifications/static/openwisp_notifications/css/notifications.css:25]: add `cursor:pointer`

File: openwisp_notifications/static/openwisp_notifications/css/notification_widget.css [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+  position: absolute;
+  z-index: 99;
+  right: 100px;
+  border: 2px solid #464646;
+  text-align: left;
+  width: 425px;
+  text-transform: none;
+  background-color: whitesmoke;
+  border-radius: 5px;
+  box-shadow: 0px 1px 20px 3px black;
```
Comment (Vivekrajput20) [inline openwisp_notifications/static/openwisp_notifications/css/notification_widget.css:11]: reduce the blur and spread in box-shadow

File: run-qa-checks [hunk 8]
```diff
@@ -1,5 +1,8 @@
 #!/bin/bash
 set -e
+
+jshint ./openwisp_notifications/static/openwisp_notifications/js/*.js
+
```
Comment (nemesisdesign) [inline run-qa-checks:5]: can we also add linting for the CSS as we have in https://github.com/openwisp/OpenWISP-Website/blob/master/run-qa-checks#L6?

**Comment by pandafy** [comment]
> * Infinite scroll does not work until the user clicks on any of the notification.

I am not able to reproduce this. :sweat_smile: 

> * mark all read and mark read does not update the notification count on the bell icon

This was because web sockets were not added on this PR. 

> * If the last notification is clicked the content is opened but content is out of the window and not visible
> * Clicking inside the notification dropdown header leads to closing of the notification dropdown. This should not happen

I was not able to replicated them, but accordion has been removed. Please see  https://github.com/openwisp/openwisp-notifications/pull/57#pullrequestreview-442624134

> * If I'm on show unread only tab and I click on mark all read, it shows me all notifications after that without switching the tab. Either switch the tab to show all or don't show read notifications.

Fixed, the tab is switched now to show all notifications. :sweat: 

> * I see you have added css and js for loader, but I can't see any loader in the notifications, I thinks it's not working properly

I have changed the implementation of showing and hiding loader. It should be visible now. :smile: 

> * Current background colors for read and unread notification are not very intuitive, I think we can keep white background for read notifications and give background color to unread notifications only. 

Sounds great, I have done that.

**Comment by pandafy** [comment]
Closing this PR in favor of #57. The development will proceed along with web sockets in #57

**Pull request closed without merging**